routing/cross_mwm_road_graph.cpp

```diff
@@ -157,70 +200,123 @@ bool CrossMwmGraph::ConstructBorderCrossImpl(OutgoingCrossNode const & startNode
 
-  // Find income node.
-  IngoingCrossNode ingoingNode;
-  bool found = false;
-  auto findingFn = [&ingoingNode, &v, &found](IngoingCrossNode const & node)
+  if (isOutgoing)
   {
-                                               if (node.m_nodeId == v.toNode.node)
-                                      {
-                                                 found = true;
-                                          ingoingNode = node;
-  }
-                                             };
-  CHECK(currentContext.ForEachIngoingNodeNearPoint(v.toNode.point, findingFn), ());
-
-  CHECK(found, ());
-
-  // Find outs. Generate adjacency list.
-  currentContext.ForEachOutgoingNode([&, this](OutgoingCrossNode const & node)
-                                     {
-                                       EdgeWeight const outWeight = currentContext.GetAdjacencyCost(ingoingNode, node);
-                                       if (outWeight != kInvalidContextEdgeWeight && outWeight != 0)
+    IngoingCrossNode ingoingNode;
+    FindIngoingCrossNode(currentContext, v.toNode, ingoingNode);
+    currentContext.ForEachOutgoingNode([&, this](OutgoingCrossNode const & node)
                                        {
-                                         vector<BorderCross> const & targets = ConstructBorderCross(node, currentMapping);
-                                         for (auto const & target : targets)
+                                         EdgeWeight const outWeight = currentContext.GetAdjacencyCost(ingoingNode, node);
+                                         if (outWeight != kInvalidContextEdgeWeight && outWeight != 0)
                                          {
-                                           if (target.toNode.IsValid())
-                                           adj.emplace_back(target, outWeight);
-                                       }
-                                       }
-                                     });
+                                           vector<BorderCross> const & targets = ConstructBorderCrossByOutgoing(
+                                                 node, currentMapping);
+                                           for (auto const & target : targets)
+                                           {
+                                             if (target.toNode.IsValid())
+                                               adj.emplace_back(target, outWeight);
+                                           }
+                                         }
+                                       });
+  }
+  else
+  {
+    OutgoingCrossNode outgoingNode;
+    FindOutgoingCrossNode(currentContext, v.fromNode, outgoingNode);
+    currentContext.ForEachIngoingNode([&, this](IngoingCrossNode const & node)
+                                      {
+                                        EdgeWeight const weight = currentContext.GetAdjacencyCost(node, outgoingNode);
```

Comment:
I can do that when I refactor code to gather OutgoingCrossNode and IngoingCrossNode which I mentioned above.